functions/access.js

```diff
@@ -0,0 +1,123 @@
+const admin = require("firebase-admin");
+const functions = require("firebase-functions");
+const { HttpsError } = require("firebase-functions/lib/providers/https");
+const { passwordsMatch } = require("./auth");
+const { uuid } = require("uuidv4");
+
+const checkIsValidToken = async (venueId, uid, token) => {
+  if (!uid) return false;
+
+  const venue = await admin.firestore().collection("venues").doc(venueId).get();
+  if (!venue.exists) {
+    throw new HttpsError("not-found", `venue ${venueId} does not exist`);
+  }
+  const granted = await venue.ref.collection("accessgranted").doc(uid).get();
+  if (!granted.exists || !granted.tokens) {
+    return false;
+  }
+
+  if (Object.keys(granted.tokens).includes(token)) {
+    // Record that the token was checked
+    if (!granted.tokens[token].usedAt) {
+      granted.tokens[token].usedAt = [];
+    }
+    granted.tokens[token].usedAt.push(new Date().getTime());
+    await venue.ref.collection("accessgranted").doc(uid).set(granted);
+    return true;
+  }
+
+  return false;
+};
+
+const getAccessDoc = async (venueId, method) => {
+  const venue = await admin.firestore().collection("venues").doc(venueId).get();
+  if (!venue.exists) {
+    throw new HttpsError("not-found", `venue ${venueId} does not exist`);
+  }
+  const accessDoc = await venue.ref.collection("access").doc(method).get();
+  return accessDoc;
+};
+
+const isValidPassword = async (venueId, password) => {
+  const access = await getAccessDoc(venueId, "password");
+  if (!access.exists || !access.data().password) {
+    return false;
+  }
+  return passwordsMatch(access.data().password, password);
+};
+
+const isValidEmail = async (venueId, email) => {
+  const access = await getAccessDoc(venueId, "emails");
+  if (!access.exists || !access.data().emails) {
+    return false;
+  }
+  return access.data().emails.includes(email.trim().toLowerCase());
+};
+
+const isValidCode = async (venueId, code) => {
+  const access = getAccessDoc(venueId, "code");
+  if (!access.exists || !access.data().codes) {
+    return false;
+  }
+  return access.data().codes.includes(code.trim());
+};
+
+const createToken = async (venueId, uid, password, email, code) => {
+  const venue = await admin.firestore().collection("venues").doc(venueId).get();
+  if (!venue.exists) {
+    throw new HttpsError("not-found", `venue ${venueId} does not exist`);
+  }
+  let granted = {};
+  let grantedDoc = await venue.ref.collection("accessgranted").doc(uid).get();
+  if (grantedDoc.exists) {
+    granted = grantedDoc.data();
+  }
+  if (!granted.tokens) {
+    granted.tokens = {};
+  }
+  // Record the first time the token was created
+  const token = uuid();
+  granted.tokens[token] = {
+    usedAt: [new Date().getTime()],
+    password: password ?? null,
+    email: email ?? null,
+    code: code ?? null,
+  };
+  await venue.ref.collection("accessgranted").doc(uid).set(granted);
```

Comment:
We can update specific fields without having to overwrite all of the data:

- https://firebase.google.com/docs/database/web/read-and-write

Any 'read then update' patterns should be done within a transaction:

- https://firebase.google.com/docs/firestore/manage-data/transactions

Reply:
I both agree and disagree with you. Transaction are mostly used for bulk writes/reads, in our case we just need to do it once. Of course overwriting the data is not very good at the moment, but we can still use .update to fix this issue. 

https://firebase.google.com/docs/database/web/read-and-write#updating_or_deleting_data

Reply:
> Transaction are mostly used for bulk writes/reads

@denisdimitrov I don't believe that is accurate.

From https://firebase.google.com/docs/firestore/manage-data/transactions:

> There are two types of atomic operations in Cloud Firestore:
> - **Transactions:** a transaction is a set of read and write operations on one or more documents.
> - **Batched Writes:** a batched write is a set of write operations on one or more documents.

Your comment appears to be referring to the latter part RE: batched writes, which isn't what I was referring to.

I was specifically referring to transactions, and their use to combine `get` and `update`/etc within a single atomic action, as described here:

- https://firebase.google.com/docs/firestore/manage-data/transactions#transactions
  - > Transactions are useful when you want to update a field's value based on its current value, or the value of some other field.
  - > A transaction consists of any number of get() operations followed by any number of write operations such as set(), update(), or delete(). **In the case of a concurrent edit, Cloud Firestore runs the entire transaction again.** For example, if a transaction reads documents and another client modifies any of those documents, Cloud Firestore retries the transaction. **This feature ensures that the transaction runs on up-to-date and consistent data.**

---

https://firebase.google.com/docs/database/web/read-and-write#updating_or_deleting_data  is useful if we are only needing to update/overwrite a single key, without using any data that already existed in firebase, and is definitely preferred over our current patterns of overwriting the entire object.

But if we need to read some data from firebase, then update a specific key, we will still need to do this in a transaction to make it atomic.

See also: https://firebase.google.com/docs/database/web/read-and-write#save_data_as_transactions

- > You can give this operation an update function and an optional completion callback. The update function takes the current state of the data as an argument and returns the new desired state you would like to write. If another client writes to the location before your new value is successfully written, your update function is called again with the new current value, and the write is retried.

Reply:
Backlogged the above as a cleanup task to fix this across the rest of the codebase in https://github.com/sparkletown/internal-sparkle-issues/issues/180 + comment to add this to our documentation in https://github.com/sparkletown/sparkle/pull/1062#issuecomment-762596026

Reply:
This is now using update, no need for transaction I suppose?

Reply:
I think if line 89 ([ref](https://github.com/sparkletown/sparkle/pull/1016/files#diff-d80040e46ddb019f6f242780e46d79059174e4fa390518d8723611c9e55aaddaR87)) is changed to be a ref, instead of using `.get()` 

-  `const grantedDocRef = await venue.ref.collection("accessgranted").doc(uid);`

And then line 97/100 use `grantedDocRef` instead of `grantedDoc`.

Then it doesn't look like we're ever actually reading from the DB, so a transaction wouldn't be necessary.

I'm not sure if the suggestions above about using the `grantedDocRef` will work with the code as is, you'd have to try it.

If we make no changes to how it currently is, then we'll still need to wrap it in a transaction, as currently there is a potential race condition between when the `grantedDoc` is read on line 87 (and whether it existed at that point), and when it is `.set()` on line 100. (Eg. if it didn't exist when line 87 ran, then it was created, then line 100 runs, we will overwrite the data that was in it, which we don't want to do)